# Voice of the Customer (VoC) Intelligence Pipeline
## [Module 1.0] — Text Preprocessing & Data Infrastructure Initialization

## Executive Context
* **Objective:** To establish a centralized file path architecture and convert unstructured, noisy user reviews into a standardized, machine-readable format for downstream sentiment modeling.
* **Input Data:** `data/raw/google_play_reviews.csv` (Original unedited app reviews).
* **Output Asset:** `data/processed/cleaned_reviews.csv` (Normalized text dataset stripped of non-semantic noise).

## Technical Stack & Dependencies
* **Core Libraries:** `pandas`, `numpy`, `re`, `nltk`
* **Methodology:** Centralized configuration management, RegEx text normalization, tokenization, stopword filtration.

---

## Pipeline Index
1. **Step 1:** Environment Initialization & Config Path Verification
2. **Step 2:** Natural Language Processing (NLP) Environment Setup
3. **Step 3:** Raw Data Ingestion & Data Integrity Audit
4. **Step 4:** The Text Normalization Engine (Regex Cleaning Pipeline)
5. **Step 5:** Processed Dataset Export & Next Pipeline Handoff

---



### Step 1: Environment Initialization & Config Path Verification

* **What we are doing:** Dynamically registering the project's root directory into the system runtime and executing an automated safety audit of our file infrastructure.
* **Why we are doing this:** Hardcoding relative file paths (e.g., `../data/file.csv`) creates fragile code that breaks when files are moved or run from different environments. By utilizing a centralized `utils/config.py` configuration layout, we establish an immutable "Source of Truth" for path management, ensuring absolute portability across different operating systems (Mac, Windows, Linux).
* **How we are doing this:** We programmatically scale up the directory tree using Python’s `pathlib` module, append the project root to `sys.path`, and run an initialization module that checks for necessary directories and builds them if they are missing.
* **Results:** A verified, structurally sound repository framework ready for safe data ingestion.

---

#### ⚙️ Script Blueprint & Data Routing
* **Execution Logic:** 
  1. Captures the current working parent directory and appends it to the system environment paths so the localized `utils` folder can be imported globally.
  2. Imports the custom `paths` manifest and the `initialize_folders` schema from your configuration layer.
  3. Executes an automated system check to visually confirm your active workspace folders.
* **Output Format & Destination:** System console log output verifying directory existence (printed directly below the cell execution block).

In [1]:
import sys
from pathlib import Path

# 1. Add project root to sys.path to access the 'utils' folder
root = Path.cwd().parent
if str(root) not in sys.path:
    sys.path.append(str(root))

# 2. Import the path manifest and the initializer
from utils.config import paths, initialize_folders

# 3. Execute the safety audit
initialize_folders()

--- Project Structure Audit ---
✅ Exists:  data/raw
✅ Exists:  data/processed
✅ Exists:  data/analysis
✅ Exists:  data/archive
✅ Exists:  reports/figures/00_sentiment_metrics
✅ Exists:  reports/figures/01_positive_analysis
📁 Created: reports/figures/02_negative_analysis


### Step 2: Natural Language Processing (NLP) Environment Setup

* **What we are doing:** Loading text-processing libraries and configuring a localized lexicon of English stopwords.
* **Why we are doing this:** General text data contains thousands of structural, non-semantic filler words (e.g., "the", "is", "at") that do not carry sentimental weight or emotional intent. To prevent these high-frequency words from drowning out actual user feedback signals during modeling, we isolate them so they can be stripped out of the dataset systematically.
* **How we are doing this:** We import `nltk` (Natural Language Toolkit), programmatically download its official English stopword dictionary corpus, and convert it into a Python `set` structure for ultra-fast, optimized lookup checks across thousands of rows.
* **Results:** A high-speed linguistic filtering array initialized and ready to clean raw strings.

---

#### ⚙️ Script Blueprint & Data Routing
* **Execution Logic:** 
  1. Imports standard data handling packages (`pandas`, `numpy`) alongside text manipulation modules (`re`).
  2. Pulls the latest NLTK `stopwords` payload directly to your local user directory.
  3. Builds an evaluation namespace using `set(stopwords.words('english'))` to maximize execution efficiency during our row-by-row matching loops.
* **Output Format & Destination:** System console string printout ("NLP Environment Ready.") verifying operational readiness.

In [6]:
import pandas as pd
import numpy as np
import re
import nltk
from nltk.corpus import stopwords

# Download the 'dictionary' of words to ignore (like 'the', 'is', 'at')
nltk.download('stopwords')
stop_words = set(stopwords.words('english'))

print("NLP Environment Ready.")

NLP Environment Ready.


[nltk_data] Downloading package stopwords to
[nltk_data]     /Users/unnatijagdishmistry/nltk_data...
[nltk_data]   Package stopwords is already up-to-date!


### Step 3: Raw Data Ingestion & Data Integrity Audit

* **What we are doing:** Loading the primary unedited dataset from the repository storage and running a programmatic structural audit to isolate and eliminate empty user feedback inputs.
* **Why we are doing this:** In Quantitative UX Research, running text analytics pipelines on rows that contain no text strings introduces fatal script errors and artificially skews sample sizes. Performing a defensive validation audit at the ingestion boundary ensures that every single entry entering the pipeline contains viable text data to be analyzed.
* **How we are doing this:** We use `pandas` to read the raw review file using our configuration path, parsing only the specific feature columns needed for our research layout (`content`, `score`, `thumbsUpCount`, `at`). We then calculate a missing value matrix, print a summary log, and systematically drop rows with empty contents using a strict research protocol.
* **Results:** A clean, optimized in-memory Dataframe with zero missing entries in our primary qualitative text targets.

---

#### ⚙️ Script Blueprint & Data Routing
* **Execution Logic:** 
  1. Pulls file configurations from `paths.DATA_RAW_GOOGLE_REVIEWS_CSV` to ingest the dataset securely.
  2. Isolates four high-value columns (`content` for text processing, `score` for rating evaluation, `thumbsUpCount` for impact scale, and `at` for temporal tracing).
  3. Uses `.isnull().sum()` to pinpoint exactly how many rows are missing actual user feedback strings.
  4. Drops null text rows via `.dropna(subset=['content'])` to prevent runtime crashes during modeling.
* **Output Format & Destination:** In-memory `pandas` DataFrame containing your filtered records, alongside system console printouts showing total row counts and null counts.

In [7]:
# Constants for pathing
#RAW_PATH = "../data/raw/google_play_reviews.csv"

# Load data - focusing on the columns we actually need for UX Research
df = pd.read_csv(paths.DATA_RAW_GOOGLE_REVIEWS_CSV, usecols=['content', 'score', 'thumbsUpCount', 'at'])

# Audit: How many reviews are missing text?
print(f"Total entries: {len(df)}")
null_counts = df['content'].isnull().sum()
print(f"Reviews missing text: {null_counts}")

# Drop rows where content is empty (Senior Protocol: Clean before processing)
df = df.dropna(subset=['content'])

Total entries: 12495
Reviews missing text: 0


### Step 4: The Text Normalization Engine (Regex Cleaning Pipeline)

* **What we are doing:** Designing and executing a scalable text-cleaning function to normalize raw review vocabulary and optimize word-matching efficiency.
* **Why we are doing this:** Mobile app reviews are uniquely chaotic, containing random uppercase shouting ("LOVE IT"), excessive punctuation ("!!!"), emojis, digits, and spelling variations. If left uncleaned, a machine learning model treats "app", "App", and "app!" as three entirely separate features, which dilutes your dataset's signals. Text normalization standardizes the lexicon so the model can capture the true voice of the customer accurately.
* **How we are doing this:** We construct a modular Python function (`clean_ux_text`) that handles three sequential tasks: casting strings to absolute lowercase, applying Regular Expressions (`re.sub`) to strip out digits and special characters while preserving standard spacing, and splitting strings into individual tokens to cross-reference and delete our pre-configured NLTK stopword array.
* **Results:** A standardized, noise-free text feature column named `cleaned_content` added directly alongside your original raw reviews for audit transparency.

---

#### ⚙️ Script Blueprint & Data Routing
* **Execution Logic:** 
  1. Declares `clean_ux_text(text)` as a robust, reusable mapping wrapper.
  2. Uses `re.sub(r'[^a-z\s]', '', text)` to matches anything that is *not* a lowercase English letter or standard whitespace, scrubbing out numbers and special characters instantly.
  3. Uses a list comprehension loop `[w for w in words if w not in stop_words]` to perform high-speed row-by-row stopword filtering.
  4. Applies the function across the entire corpus using `.apply()` and displays a comparative snapshot via `.head()`.
* **Output Format & Destination:** A modified in-memory DataFrame displaying a side-by-side view of the original `content` and the new `cleaned_content` structures.

In [8]:
def clean_ux_text(text):
    """
    Standardizes review text for NLP analysis.
    - Lowers case
    - Removes special characters/digits
    - Removes stopwords
    """
    # 1. Lowercase
    text = str(text).lower()
    
    # 2. Remove special characters and digits using Regular Expressions (Regex)
    text = re.sub(r'[^a-z\s]', '', text)
    
    # 3. Tokenize (split into words) and remove stopwords
    words = text.split()
    cleaned_words = [w for w in words if w not in stop_words]
    
    return " ".join(cleaned_words)

# Apply the function to our content column
# Warning: This may take a minute if the dataset is 100k+ rows
print("Cleaning text... please wait.")
df['cleaned_content'] = df['content'].apply(clean_ux_text)

print("Success: Text normalization complete.")
df[['content', 'cleaned_content']].head()

Cleaning text... please wait.
Success: Text normalization complete.


,content,cleaned_content
0,I cannot open the app anymore,cannot open app anymore
1,I have been begging for a refund from this app...,begging refund app month nobody replying
2,Very costly for the premium version (approx In...,costly premium version approx indian rupees pe...
3,"Used to keep me organized, but all the 2020 UP...",used keep organized updates made mess things c...
4,Dan Birthday Oct 28,dan birthday oct


### Step 5: Processed Dataset Export & Next Pipeline Handoff

* **What we are doing:** Isolating our finalized research features and writing the standardized DataFrame into permanent disk storage.
* **Why we are doing this:** To maintain clear tracking of our data assets and ensure clean boundaries between project phases, we isolate our heavy text normalization work from our analytical phases. Exporting this intermediate data asset means downstream modeling notebooks can instantly ingest clean text strings without recalculating regex patterns or reloading NLTK objects every single time.
* **How we are doing this:** We slice our in-memory DataFrame to retain only the essential data columns (`content`, `cleaned_content`, `score`, `at`, `thumbsUpCount`). We then invoke the `pandas.to_csv` function, mapping it to our configuration-driven path manifest with row indexes disabled to preserve structural integrity.
* **Results:** A clean, optimized `.csv` file saved directly inside your local `data/processed/` directory, establishing the foundational dataset for Module 2.0.

---

#### ⚙️ Script Blueprint & Data Routing
* **Execution Logic:** 
  1. Filters the runtime DataFrame (`df_to_save`) to discard unnecessary metadata.
  2. References `paths.DATA_PROCESSED_CLEANED_CSV` from our centralized configuration layer to identify the exact storage location.
  3. Writes the clean records to disk using `index=False` to avoid generating trailing artifact index columns.
* **Output Format & Destination:** A physical CSV asset generated at `data/processed/cleaned_reviews.csv`, along with a console string log verifying successful execution.

In [9]:
# Save the cleaned dataframe to the processed folder
# We keep 'content' for reference and 'score' for validation
#PROCESSED_PATH = "../data/processed/cleaned_reviews.csv"

df_to_save = df[['content', 'cleaned_content', 'score', 'at', 'thumbsUpCount']]
df_to_save.to_csv(paths.DATA_PROCESSED_CLEANED_CSV, index=False)

print(f"Data Cleaning Complete. File saved at: {paths.DATA_PROCESSED_CLEANED_CSV}")

Data Cleaning Complete. File saved at: /Users/unnatijagdishmistry/Documents/UX_Mixed-Methods_Research_Projects/sentiment-analysis-audit/data/processed/cleaned_reviews.csv


## Module 1.0 Summary & Downstream Handoff

### Phase Evaluation
Through this initial data engineering phase, we successfully established a robust, configuration-driven foundation for our text analytics pipeline. By migrating away from hardcoded paths and setting up an immutable path manifest (`utils/config.py`), this repository satisfies standard, production-ready portability requirements.

The raw mobile review text data was successfully audited, eliminating blank feedback instances that could introduce downstream valuation skewing or execution errors. Through our text normalization engine, we handled the inherent noise of mobile marketplace feedback (irregular capitalization, digits, special characters, and non-semantic linguistic filler). This process significantly increased our qualitative data's overall signal-to-noise ratio.

### Pipeline Metrics At-A-Glance
* **Ingested Records:** 12,495 entries processed from raw storage.
* **Data Integrity Checks:** 0 null-text rows detected and removed during the ingestion audit.
* **Feature Engineering:** `cleaned_content` created as a tokenized, noise-filtered, standardized target text array.

---

### Next Pipeline Handoff: Proceed to Module 2.0
The resulting file asset—`data/processed/cleaned_reviews.csv`—is now physically written to disk. This file contains the foundational structured text required for our modeling work. 

In the next phase, **Module 2.0 (Sentiment Modeling & Cognitive Dissonance Analysis)**, we will ingest this clean dataset to build out our automated sentiment layer, cross-verify model confidence boundaries, and mathematically isolate instances where user star-ratings diverge from their written text expression.